# Basics of Python


## Boolean Variables

In [65]:
str = "SumarFawaz123"

In [69]:
print(str.isalnum()) #returns true if the variable has numerics and alphabets but shouldn't have a space between words
print(str.isalpha()) #returns true if the variable has only strings
print(str.isdigit()) #returns true if the variable has digits
print(str.istitle()) #returns true if the the variable has title words [ first character in every word is upper case]
print(str.isupper()) #returns true if all the words are in upper case
print(str.islower()) #returns true if all the words are in lower case
print(str.isspace()) #returns true if there are spaces in the word
print(str.endswith('z')) #returns true if the word ends with the parsed letter (Case Sensitive)
print(str.startswith('S')) #returns true if the word starts with the parsed letter (Case Sensitive)

True
False
False
False
False
False
False
False
True


## Boolean and Logical Operators

In [72]:
True and True

True

In [74]:
True and False

False

In [76]:
True or False

True

In [78]:
True or True

True

In [80]:
False or False

False

## Lists in Python
- Lists are defined within square brackets [] which are changeable, mutable and can contain a mixture of data types. A list can be written using Commas Seperated Values. List is a class in Python.

In [83]:
type([])

list

In [1]:
lst_example = [] # First way of defining a list
print(type(lst_example))

lst = list() # Inbuilt way of using list() method to create a list
print(type(lst))

lst = ["English","Sumar",100,1000,10000] # Parsing Values into List
print(lst) # Print all values in a list just by using print() method 
lst # or just call it

<class 'list'>
<class 'list'>
['English', 'Sumar', 100, 1000, 10000]


['English', 'Sumar', 100, 1000, 10000]

In [115]:
len(lst) # Prints the number of elements inside list or any other variable

print("Total Length of the list {length}".format(length=len(lst)))

Total Length of the list 5


### Inbuilt Functions in List

In [118]:
lst.append("Sumar added to the end of the list") #append adds elements to the end of the list

In [122]:
print(lst)

['English', 'Sumar', 100, 1000, 10000, 'Sumar added to the end of the list']


In [124]:
lst[1] # Accessing specific elements from the List

'Sumar'

In [130]:
lst[2:] # Accessing elements in the List from a Range lst[from index:last index]

[100, 1000, 10000, 'Sumar added to the end of the list']

In [134]:
lst[1:3]

['Sumar', 100]

In [136]:
lst.append(["Hello", "World"]) # Creating Nested Lists

In [138]:
print(lst)

['English', 'Sumar', 100, 1000, 10000, 'Sumar added to the end of the list', ['Hello', 'World']]


### Insert
- Adds Elements to Specific Index in the List

lst.insert(0,"Added to front of list") # Used to enter elements to specific indexes

In [143]:
lst

['Added to front of list',
 'English',
 'Sumar',
 100,
 1000,
 10000,
 'Sumar added to the end of the list',
 ['Hello', 'World']]

### Extend 
- Adds elements to the end of the list without nesting with multiple values

In [152]:
lst.extend([10,15]) # Parsing values inside curly braces

In [156]:
lst

['Added to front of list',
 'English',
 'Sumar',
 100,
 1000,
 10000,
 'Sumar added to the end of the list',
 ['Hello', 'World'],
 10,
 15,
 10,
 15]

### Operations on Lists

In [159]:
lst = [1,2,3,4,5,6]

In [161]:
sum(lst)

21

### Pop() Method
- Removes the Last element by default if not specified

In [164]:
lst.pop()

6

In [169]:
lst.pop(0) # Removes 0th element

2

In [171]:
lst

[3, 4, 5]

In [175]:
lst.extend([5])

In [177]:
lst

[3, 4, 5, 5]

### Counts total occurence of a given element within the list

In [180]:
lst.count(5)

2

### index() method

In [187]:
lst.index(4,0,3) # Finds the index of the parsed element which is the first parameter, 
                 #2nd and 3rd parameters are the start and end elements in the list

1

### min() and max() methods

In [194]:
min(lst) # Lowest Value in the list

3

In [196]:
max(lst) # Highest Value in the list

5

## SETS
- A Set is an unordered collection data type that is iterable, mutable, and has no duplicate elements. Python's set class is build upon the fundamentals of mathematical set theory. It is build using the Hash Table data structure.

In [2]:
## Defining an empty Set

set_var = set()
print(set())
print(type(set_var))

set()
<class 'set'>


In [4]:
# Set in Python ignores or removes duplicates 
set_var = {1,2,3,4,3}

In [6]:
print(set_var)

{1, 2, 3, 4}


### Inbuilt Functions In Sets

In [9]:
## Inbuilt function for adding values to the set 
set_var.add("Hulk")
print(set_var)

{1, 2, 3, 4, 'Hulk'}


In [11]:
## Set Functions 

set1={"Avengers","Ironman","Hitman"}
set2={"Avengers","Ironman","Hitman","Hulk2"}

In [15]:
# intersection function - Gets the common elements in both the sets 
set2.intersection(set1)

{'Avengers', 'Hitman', 'Ironman'}

In [21]:
# intersection.update function - Gets rid of the element that is not common for both
set2.intersection_update(set1)

In [29]:
set2

{'Avengers', 'Hitman', 'Ironman'}

In [31]:
# Difference Function - Gets the value that is not common in both the sets
set2.difference(set1)

set()

In [33]:
set2

{'Avengers', 'Hitman', 'Ironman'}

## DICTIONARIES
- A dictionary is a collection which is unordered, changeable and indexed. In python dictionaries are written with curly brackers, and they have keys and values.

In [37]:
# Defining Dictionary using {} 
dic = {}

In [39]:
type(dic)

dict

In [41]:
## Creating a dictionary 

## Creating a dictionary is possible with the inbuilt dict() function 

my_dictionary = {"Car1":"Audi","Car2":"BMW","Car3":"Benz"}

In [43]:
type(my_dictionary)

dict

In [45]:
## Accessing item values based on keys 
my_dictionary['Car1']

'Audi'

In [51]:
## We Can Loop through the dictionaries keys 
for x in my_dictionary:
    print(x)
    print(my_dictionary[x])

Car1
Audi
Car2
BMW
Car3
Benz


In [53]:
## Directly Looping through the dictionaries values 
for x in my_dictionary.values():
    print(x)

Audi
BMW
Benz


In [55]:
## Looping through all the items in the dictionary 
for x in my_dictionary.items():
    print(x)

('Car1', 'Audi')
('Car2', 'BMW')
('Car3', 'Benz')


In [57]:
## Adding items in the dictionary 
my_dictionary['Car4']='Pagani'

In [59]:
my_dictionary

{'Car1': 'Audi', 'Car2': 'BMW', 'Car3': 'Benz', 'Car4': 'Pagani'}

In [65]:
## Overridden Key Values in dictionary 
my_dictionary['Car1']='Ferari'

In [67]:
my_dictionary

{'Car1': 'Ferari', 'Car2': 'BMW', 'Car3': 'Benz', 'Car4': 'Pagani'}

### Nested Dictionary 

In [1]:
## Creating key values and storing them in a dictionary with a key

car1_model = {'Mercedes':1960}
car2_model = {'Audi':1970}
car3_model = {'Ambassador':1980}

car_type = {'car1':car1_model,'car2':car2_model,'car3':car3_model}

In [3]:
car_type

{'car1': {'Mercedes': 1960},
 'car2': {'Audi': 1970},
 'car3': {'Ambassador': 1980}}

In [5]:
## Accessing Items in the dictionary 
print(car_type['car1'])

{'Mercedes': 1960}


In [7]:
## Accessing items one step further from the nested key 
print(car_type['car1']['Mercedes'])

1960


### Tuples
- Tuples are not mutable or changeable. The elements cannot be changed

In [15]:
## creating a tuple using tuple() function 

my_tuple = tuple()

In [17]:
type(my_tuple)

tuple

In [19]:
## Another way of creating tuple using ()
my_tuple = ()


In [21]:
type(my_tuple)

tuple

In [23]:
my_tuple = ("Krish","Ankur","John")

In [27]:
## Accessing elements in a tuple
my_tuple[0]

'Krish'

In [29]:
## Counting the number of occurences of an item 

my_tuple.count('Krish')

1

# Lambda Function 
Lambda Function is an anonymous function which has no name is faster than traditional function

In [2]:
# A Normal Function 

In [4]:
def addition(a,b):
    return a+b

In [6]:
addition(4,5)

9

In [14]:
# Lambda Function 
#Works only with single expressions

In [16]:
addition_lambda = lambda a,b:a+b

In [18]:
addition_lambda(2,4)

6

# Map Function

In [21]:
# Map Function is used for iterating through the whole loop without using a for loop 

In [23]:
def even_or_odd(num):
    if num%2==0:
        return "The number {} is even".format(num)
    else:
        return "The number {} is odd".format(num)

In [25]:
even_or_odd(24)

'The number 24 is even'

In [27]:
lst = [2,3,4,5,6,7,8,9]

In [31]:
list(map(even_or_odd,lst))

['The number 2 is even',
 'The number 3 is odd',
 'The number 4 is even',
 'The number 5 is odd',
 'The number 6 is even',
 'The number 7 is odd',
 'The number 8 is even',
 'The number 9 is odd']

# Filter Function
Returns all the values based on the condition without returning bool values

In [37]:
def even(num):
    if num%2 == 0:
        return True

In [39]:
lst = [2,3,4,5,6,7,8,9,0]

In [47]:
list(filter(even,lst)) # function name and list variable

[2, 4, 6, 8, 0]

In [49]:
# Returns all the even numbers from the list without saying true or false

In [51]:
# Using lambda in this scenario to avoid the function

In [53]:
list(filter(lambda num:num%2==0, lst))

[2, 4, 6, 8, 0]

In [57]:
#short hand of using functions with one expression in the above example using lambda

In [59]:
list(map(lambda num:num%2==0, lst))

[True, False, True, False, True, False, True, False, True]

In [61]:
#map returns all the values with true or false based on the condition

# List Comprehension
List comprehensions provides a concise way to create lists. It consists of brackets containing an expression followed by a for clause, then zero or more for or if clauses. The expressions can be anything, meaning you can put in all kinds of objects in lists

In [64]:
lst1=[]
def lst_square(lst):
    for i in lst:
        lst1.append(i*i)
    return lst1

In [66]:
lst_square([1,2,3,4,5,6,7,8,9])

[1, 4, 9, 16, 25, 36, 49, 64, 81]

In [68]:
#Returns all the squares of numbers in the list 

In [70]:
#Using List Comprehension can get this done in one line of code

In [78]:
#list comprehension
#[returning value or expression | for clause]
lst = [1,2,3,4,5,6]
[i*i for i in lst]
# the whole function is shortened to one line of code

[1, 4, 9, 16, 25, 36]

In [80]:
# we can also parse it to a variable
lst1 = [i*i for i in lst]

In [82]:
lst1

[1, 4, 9, 16, 25, 36]

In [88]:
# adding conditions to list comprehension
lst1 = [i*i for i in lst if i%2==0]
# gets the square of even numbers only

In [90]:
lst1

[4, 16, 36]

# String Formatting in Python

In [101]:
firstname = "Sumar"
lastname = "Fawaz"

fullname = "Hello your first name is {first_name} and your last name is {last_name}".format(first_name=firstname,last_name=lastname)

In [103]:
fullname

'Hello your first name is Sumar and your last name is Fawaz'

In [105]:
fullname = "Hello your first name is {} and your last name is {}".format(firstname,lastname)

In [107]:
fullname

'Hello your first name is Sumar and your last name is Fawaz'

# Python List Iterables vs Iterators
Iterables will be initiated in the memory whenever initilialized while iterators won't be initiated in the memory until the .next() will be used.

- An iterator is very useful to write code which performs better because all the values are not stored inside the memory until it is being called

In [161]:
#Lists are iterables
lst = [1,2,4,5,6,7,8,9,0]

In [163]:
for i in lst:
    print(i)

1
2
4
5
6
7
8
9
0


In [175]:
lst1 = iter(lst) #inbuilt iter() method, this is an iterator now. Convert iterable to iterator

In [177]:
# The elements in the iterator cannot be seen at once, instead we shoudl use the method next()
# to get each and every element in the iterator

In [179]:
next(lst1)

1

In [181]:
# WE can write a for loop to get the values inside the iterator

In [183]:
for i in lst1:
    print(i)

2
4
5
6
7
8
9
0


In [185]:
# it is skipping 1 because 1 is already stored in the memory

# OOPS in Python 

## Class in Python

In [215]:
class Car:
    pass #creating a class and saying it is empty

In [217]:
car1 = Car()

In [219]:
car1

In [221]:
car1.windows=5
car1.doors=4

In [225]:
print(car1.windows)

5


In [77]:
class Car:
    def __init__(self,window,door,enginetype):
        self.windows=window
        self.doors=door
        self.enginetype=enginetype
# __init__ method is a constructor for the class where it will be initializaed each and everytime
# the class is being called
    def self_driving(self):
        return "This car has a {engine_type} engine".format(engine_type=self.enginetype)
        # This is a method that we have created called self_driving() where we have parse the 
        # parameter self inside the method. We can use that to access the values in the above method
        # we have to use the self keyword infront of the variable that we want to call

In [79]:
# When we try to create an object for that class we have to parse the values for the init method 
car1=Car(4,4,"Diesel")

In [81]:
car1.windows
# We have to use the self init values to access the variables not the parameter elements

4

In [83]:
car1.self_driving() # calling that method here 

'This car has a Diesel engine'

# Python Exception Handling

In [86]:
#Handling Exceptions can be very specific and generic. We can use the general class that is called 
# Exception which we can use to handle exceptions in a generic way

#Python has try except not try catch 

try:
    #code we want to execute
    a=b
except:
    print("Error Exception")

Error Exception


In [88]:
# In the above code there is an exception that has occured while executing the code base. so the 
# except has captured that and is a given a custom message

In [90]:
try:
    a=b
except Exception as ex:
    print(ex)

name 'b' is not defined


In [92]:
# In the above code snippet we have the except which handles the exception. For that we are parsing
# the generic Exception class and we have an alias called ex that we are printing to show the 
# exception 

In [94]:
# We can capture more precise exceptions based on their type that is printed during the exception 

In [96]:
a=b

NameError: name 'b' is not defined

In [98]:
# in the above code snippet we haven't used a try except so we are getting this exception unhandled
# directly to the user which is pretty much hard to understand and is scary
# we can handle this using the subclass or child class of the main class which is Exception

In [100]:
try:
    a=b
except NameError:
    print("You haven't initiated the name b")
except Exception as ex:
    print(ex)

You haven't initiated the name b


In [102]:
# In the above code snippet we are using specific exception handling at the top and we also have the
# generic exception handling below that to handle exceptions that we haven't foreseen

In [122]:
try:
    a=int(input("Please enter number 1"))
    b=int(input("Please enter number 2"))
    c=int(a/b)
    
    #We are trying to divide two values, we might get errors in terms of dividing it using 0
    #Which is an exception
except ZeroDivisionError:
    print("You can't divide a value by 0, please enter a value greater than 0.")
    
except Exception as ex:
    print("Something just went wrong, please contact the support team.")

else:
    # If the division is successful, i will be displaying the result
    print(c)

finally:
    print("The calculation was performed")
    # The code snippet within finally will execture regardless of success or failure of the try block

# In python we have else and finally that we can add to the try except block. 
# The else will only execute if the except has not caught any exception. 
# finally will run even if there are exceptions or not

# the main feature of else is that we can also access local variables inside try block
# which cannot be done in other programming languages


Please enter number 1 12
Please enter number 2 0


You can't divide a value by 0, please enter a value greater than 0.
The calculation was performed


In [124]:
12/0 # ZeroDivisionError

ZeroDivisionError: division by zero

# Custom Exception 

In [127]:
# We can create our custom exceptions that will handle different use case scenarios within the 
# program
# We will be creating a generic class where we will parse the Exception class as an arguement
class Error(Exception):
    pass

class dobException(Error):
    pass
# We created the generic class called Error above the dobException class which is a custom 
# Exception that we are trying to handle. WE have parsed the Error class that we have created above 
# that method

In [133]:
year = int(input("Please enter the year of birth"))
age = 2024-year
#we will write the code snipped within the try except block 
try:
    if age<=30 and age>20:
        print("You are eligible for this")
    else:
        raise dobException # now for the else condition we are raising the dobException that we 
        #created
except dobException:
    # Now we have used the same class to parse if for the except as well
    print("The age is not within the range")

Please enter the year of birth 1999


You are eligible for this


# Inheritance in Python 

In [136]:
class Car:
    def __init__(self,window,door,enginetype):
        self.windows=window
        self.doors=door
        self.engtype=enginetype
    def drive(self):
        print("This person drivess")

In [156]:
# Inheritance of the above Car class
class audi(Car): # this is where we are inherting the Car class
    def __init__(self,windows,doors,engtype,ai): # ai is additional parameter here
        super().__init__(windows,doors,engtype) # here super() refers to the parent class Car and
        # we should use the variables that we used with self in the parent class
        self.enableai=ai # This refers to the additional parameter that we parsed
    def selfdriving(self):
        print("Audi Supports self driving")

In [158]:
audiA6 = audi(4,5,"Petrol",True)

In [160]:
audiA6.windows

4

In [162]:
audiA6.selfdriving()

Audi Supports self driving


In [164]:
# Accessing drive() method in the Car class

In [166]:
audiA6.drive()

This person drivess


# OOPS-Magic Methods in Classes 

In [181]:
# Magic methods are the methods in a class that run internally to do certain functionalities
# in the class 

In [183]:
dir(Car)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'drive']

In [179]:
# The above displayed methods are the magic methods inside the Car class

In [195]:
# These magic methods can be overriden just like the __init__ method that we override

In [197]:
c=Car(5,4,'Petrol')

In [199]:
print(c.__sizeof__())

16


In [201]:
# We have used a magic method above which is __sizeof__(). Now we will be overriding it

In [209]:
# Creating a new class to override the __sizeof__() method
class Human:
    def __init__(self,name):
        self.hname=name
    def __sizeof__(self):
        return "Overriding the magic method"

In [211]:
human = Human("Sumar")

In [213]:
human.__sizeof__()

'Overriding the magic method'

In [215]:
# that is how magic methods work in python

# Python Assert Statement
Assert statement checks if a given logical condition is true or false. Program execution proceeds only if the expression is true and raises the the AssertionError when it is false. 

In [219]:
# Assert can be used instead of an if because when we want to only proceed if it is true
# we can handle the else statement or the faiure of the logical expression using Assertion exception
# that is thrown 

In [221]:
num=10 
assert num>=10

In [223]:
# now we are not getting any exception because it is true

In [225]:
num = 10 
assert num>10

AssertionError: 

In [227]:
#now we are getting the AssertionError exception which we can use 

In [233]:
try:
    num=int(input("Please enter the value"))
    assert num%2==0
    print("The number is even")
except AssertionError:
    print("The number is not an even number")

Please enter the value 13


The number is not an even number


In [237]:
# We have used the AssertionError to handle the false of the condition

# Iterators Vs Generators

In [269]:
# Iterator
lst = [1,2,3,4,5]

In [271]:
iterable = iter(lst)

In [273]:
next(iterable)

1

In [275]:
### Generator 
def square(n):
    for i in range(n): #eturns a sequence of numbers, starting from 0 by default, 
        #and increments by 1 (by default), 
        #and stops before a specified number
        yield i**2

### Difference between Iterator & Generator
1. To create an iterator, we should use the iter() method and to create a generator we should create a function with the keyword yield just like the above example
2. The yield keyword can save the local variable value and also print it.
3. Generators help us write fast and compact code
4. Iterator is much more memory efficient compared to Generator

### Decorators

In [299]:
## Function copy
    ### When we assign a function to a variable and use del keyword to delet the function and 
    ### still try to access the function using the variable. it will still work
    ### which means the function is getting copied to the variable completely
## Closures
    ### Closures are basically function within a function
## Decorators

In [301]:
def welcome():
    return "Welcome"

In [303]:
wel=welcome() # this line of code basically copies the function welcome() to variable wel
del welcome # now we are deleting the function from the memory

In [305]:
wel #now if we use the variable it will still do the functionalities inside a function

'Welcome'

In [307]:
welcome() # This will give an exception saying it is not defined, means doesn't exist

NameError: name 'welcome' is not defined

In [315]:
### Closures 
def main_welcome():
    msg = "Hello Sumar"
    def welcome_message():
        print(msg)
        print("we can access variables of the parent function from the inner function")
    return welcome_message()
    

In [317]:
## now when we call the main_welcome() function we will get the welcome_message() functions because
## we have returned the welcome_message() function

In [319]:
main_welcome()

Hello Sumar
we can access variables of the parent function from the inner function


In [333]:
### Closures but initial Decorators 
### We can parse functions as a parameter which is the starting point of decorators
def main_welcome(func):
    msg = "Hello Sumar"
    def welcome_message():
        func("This is a function that is getting printed")
        print("we can access variables of the parent function from the inner function")
    return welcome_message()
    

In [335]:
main_welcome(print) # we are basically parsing the inbuilt print function

This is a function that is getting printed
we can access variables of the parent function from the inner function


In [353]:
### Decorators
### Parsing a custom function inside a function is called Decorators
def main_welcome(func):
    msg = "Hello Sumar"
    def welcome_message():
        func()
        print("we can access variables of the parent function from the inner function")
    return welcome_message()
### this function takes func as a parameter where we can parse our own function. 
### It is also called inside subclass of the main class function

In [397]:
#@main_welcome # making this as a decorator where we can use it directly and it will parse the
#channel_name 
# we can also access all the things inside the decorator
main_welcome
def channel_name():
    print("This is from the channel name function")

In [399]:
# Here we will be parsing the channel_name() as a parameter in main_welcome() function 

In [401]:
main_welcome(channel_name)

This is from the channel name function
we can access variables of the parent function from the inner function


In [403]:
# The above mentioned concepts are known as Decorators

In [405]:
# we can also initialize the whole thing to a variable 
a=main_welcome(channel_name)

This is from the channel name function
we can access variables of the parent function from the inner function


## Class Methods and Class Variables

In [459]:
class Car:
    base_price=10000 ## Class Variables
    def __init__(self,windows,doors,power):
        self.windows=windows
        self.doors=doors
        self.power=power
    def what_base_price():
        print("The base price is {}".format(self.base_price))
    @classmethod # Creating class methods that can be directly accessed for certain purposes this is 
    #also a decorator
    def revise_base_price(cls,inflation):#we have to use the keyword cls inside the parameter to refer to the class
        cls.base_price=cls.base_price+cls.base_price*0.20

In [461]:
Car.base_price ## Directly accessing class variables

10000

In [463]:
Car.revise_base_price(0.20)

In [465]:
Car.base_price

12000.0

In [467]:
# this is how we can use class methods to update class variables and do other operations

# Static Method

In [484]:
import datetime
now = datetime.datetime.now()

In [508]:
class Car:
    base_price=10000 ## Class Variables
    def __init__(self,windows,doors,power):
        self.windows=windows
        self.doors=doors
        self.power=power
    def what_base_price():
        print("The base price is {}".format(self.base_price))
    @classmethod # Creating class methods that can be directly accessed for certain purposes this is 
    #also a decorator
    def revise_base_price(cls,inflation):#we have to use the keyword cls inside the parameter to refer to the class
        cls.base_price=cls.base_price+cls.base_price*0.20
    @staticmethod # static method doesnt require of parsing self or cls
    def check_year(year):
        if now.year > year:
            print("The date is behind the year")
            # calling functions inside a static function 
            Car.revise_base_price(0.20)
        else:
            print("The date is equal or after the year")

In [510]:
Car.check_year(2020)

The date is behind the year


In [512]:
# This is what the static method does, where there is no need of specifying cls or self

In [514]:
Car.base_price

12000.0

### Multiple Inheritance

In [517]:
class A:
    def method1(self):
        print("A class method is called")

In [519]:
class B(A):
    def method1(self):
        print("B class method is called")
    def method2(self):
        print("B class method2 is called")

In [523]:
class C(A):
    def method1(self):
        print("C class method is called")

In [536]:
class D(B,C): # we can use a comma and inherit more than one class
    def method1(self):
        print("D class method is called")
        C.method1(self)
        B.method1(self)
        A.method1(self)
# This is how we refer to the classes of the parent classes to call it inside the child class

In [538]:
d=D()

In [540]:
d.method1() #when we called method1, all the method1 from other inherited class is being called

D class method is called
C class method is called
B class method is called
A class method is called


In [544]:
B.method1(d) # calling the parent function class from the child class

B class method is called


In [547]:
A.method1(d)

A class method is called


### Eval() Function
Evaluates python expressions which are written as strings. This can run any function or calculation
parsed in the form of strings

In [556]:
# Eval()
eval("5*10/5")

10.0

### How does Eval() work?
1. Parse Python Expression
2. Compile into a byte code
3. Evaluate the Expression
4. Return the Result

In [560]:
# an inbuilt method called compile() is used to convert the expression into byte code

In [566]:
var = compile("5+5","<string>","eval") #this basically gives the expression as a string which is
#"5+5"
# then we are specifying that it is a string
#"<string>"
# then we are specifying which method should take care of this whether it is eval() or exec()

In [568]:
eval(var)

10

#### Globals Parameter

In [572]:
### Globals

In [574]:
x=10 
eval("x+50+x**2",{"x":x}) # Here we are writing the expression as a string and in the string 
# we have something called x which is not defined anywhere in the code, so this will throw and error
# we can parse the value of x inside the eval method as a global parameter which follows the 
# dictionary data structure

160

In [576]:
z=100
x=200 

In [578]:
eval("x+z",{"x":x,"z":z})

300

In [580]:
# We can also declare the variables the eval() will get those as variables and run it inside the 
# eval() function 
z=100 
x=200


In [582]:
eval("x+z")

300

In [584]:
# Still it will execute fine without any problem

In [586]:
# We can also perform boolean expressions

In [588]:
a=100
b=200

In [590]:
eval("b>a")

True

 ## Asynchronous Execution
 - It is basically running different threads or tasks by saving the progress of one particular task
 - This is also known as multi-threading
 - Python has two classes that can do this, one is ThreadPoolExecutor & ProgressPoolExecutor

In [599]:
import time
from concurrent.futures import ThreadPoolExecutor
def returnnumber(a):
    time.sleep(1)
    return a

In [601]:
returnnumber(12)

12

In [597]:
start=time.time() # This is basically used to calculate the time for the process
with ThreadPoolExecutor(max_workers=None) as executor: # max_workers are the number of concurrent
    #processes it should run 
    for result in executor.map(returnnumber,range(100)):
        print("Count: {0}".format(result))
print("The total time {0}".format(time.time()-start))

Count: 0
Count: 1
Count: 2
Count: 3
Count: 4
Count: 5
Count: 6
Count: 7
Count: 8
Count: 9
Count: 10
Count: 11
Count: 12
Count: 13
Count: 14
Count: 15
Count: 16
Count: 17
Count: 18
Count: 19
Count: 20
Count: 21
Count: 22
Count: 23
Count: 24
Count: 25
Count: 26
Count: 27
Count: 28
Count: 29
Count: 30
Count: 31
Count: 32
Count: 33
Count: 34
Count: 35
Count: 36
Count: 37
Count: 38
Count: 39
Count: 40
Count: 41
Count: 42
Count: 43
Count: 44
Count: 45
Count: 46
Count: 47
Count: 48
Count: 49
Count: 50
Count: 51
Count: 52
Count: 53
Count: 54
Count: 55
Count: 56
Count: 57
Count: 58
Count: 59
Count: 60
Count: 61
Count: 62
Count: 63
Count: 64
Count: 65
Count: 66
Count: 67
Count: 68
Count: 69
Count: 70
Count: 71
Count: 72
Count: 73
Count: 74
Count: 75
Count: 76
Count: 77
Count: 78
Count: 79
Count: 80
Count: 81
Count: 82
Count: 83
Count: 84
Count: 85
Count: 86
Count: 87
Count: 88
Count: 89
Count: 90
Count: 91
Count: 92
Count: 93
Count: 94
Count: 95
Count: 96
Count: 97
Count: 98
Count: 99
The total 

In [603]:
start=time.time() # This is basically used to calculate the time for the process
with ThreadPoolExecutor(max_workers=1) as executor: # max_workers are the number of concurrent
    #processes it should run 
    for result in executor.map(returnnumber,range(100)):
        print("Count: {0}".format(result))
print("The total time {0}".format(time.time()-start))
# Playing with different max_workers, more amount of max_workers will increase the speed of execution


Count: 0
Count: 1
Count: 2
Count: 3
Count: 4
Count: 5
Count: 6
Count: 7
Count: 8
Count: 9
Count: 10
Count: 11
Count: 12
Count: 13
Count: 14
Count: 15
Count: 16
Count: 17
Count: 18
Count: 19
Count: 20
Count: 21
Count: 22
Count: 23
Count: 24
Count: 25
Count: 26
Count: 27
Count: 28
Count: 29
Count: 30
Count: 31
Count: 32
Count: 33
Count: 34
Count: 35
Count: 36
Count: 37
Count: 38
Count: 39
Count: 40
Count: 41
Count: 42
Count: 43
Count: 44
Count: 45
Count: 46
Count: 47
Count: 48
Count: 49
Count: 50
Count: 51
Count: 52
Count: 53
Count: 54
Count: 55
Count: 56
Count: 57
Count: 58
Count: 59
Count: 60
Count: 61
Count: 62
Count: 63
Count: 64
Count: 65
Count: 66
Count: 67
Count: 68
Count: 69
Count: 70
Count: 71
Count: 72
Count: 73
Count: 74
Count: 75
Count: 76
Count: 77
Count: 78
Count: 79
Count: 80
Count: 81
Count: 82
Count: 83
Count: 84
Count: 85
Count: 86
Count: 87
Count: 88
Count: 89
Count: 90
Count: 91
Count: 92
Count: 93
Count: 94
Count: 95
Count: 96
Count: 97
Count: 98
Count: 99
The total 

In [604]:
# The number of execution that is doing now is 1 by 1 because the max_workers is set to 1

## Zip() Function-Parallel Iteration

In [632]:
#zip() basically does parallel iteration from different lists and creates 
#tuples from the lists as tuples whereas for an example the 0th element of the first list and the
#0th element of the second list will be made as a tuple

#we can iterate through multiple iterables

In [634]:
lst1 = [1,2,3,4]
lst2 = [4,5,6,7]

output=zip(lst1,lst2)
print(type(output))

for i,j in output:
    print(i,j)


<class 'zip'>
1 4
2 5
3 6
4 7


In [636]:
# It prints it all the values in both list by taking the respective indexes

In [638]:
zip(lst1,lst2)

In [640]:
list(zip(lst1,lst2))

[(1, 4), (2, 5), (3, 6), (4, 7)]

In [642]:
# This is how it create types of the the data of different sets matching their indexes

In [644]:
type(zip())

zip

In [646]:
print(type(lst1))

<class 'list'>


In [648]:
# We can also use zip() also with dictionaries

In [650]:
dict1={'name':'Sumar','last_name':'Fawaz','age':30}
dict2={'name':'Alpha','last_name':'Beta','age':23}

In [652]:
dictionary=zip(dict1.items(),dict2.items())

In [654]:
for (i,j),(i2,j2) in dictionary:
    print(i,j)
    print(i2,j2)
# we are basically iterating through both the items in the dict1 and dict2 through a for loop

name Sumar
name Alpha
last_name Fawaz
last_name Beta
age 30
age 23


## Logging 

In [2]:
import logging #importing logging class

In [4]:
#giving the basic txt file config

In [6]:
logging.basicConfig(filename="log.txt",filemode='a',format='%(asctime)s %(levelname)s-%(message)s',
                   datefmt='%Y-%m-%d %H:%M:%S') # this is basic file rules or config

In [8]:
for i in range(0,15):
    if i%2==0:
        logging.warning('Log Warning Message')
    elif i%3==0:
        logging.critical('Log Critical Message')
    else:
        logging.error('Log Error Message')
#here we are logging several different logs with different criticality like warning,critical and error
        

## SHA, ( Secure Hash Algorithms )- Base Of BlockChain ties.

1. SHA256 : This hash function belong to hash class SHA-2, the internal block size of it is 32 bits.

2. SHA384 : This hash function belong to hash class SHA-2, the internal block size of it is 32 bits. This is one of the truncated version.

3. SHA224 : This hash function belong to hash class SHA-2, the internal block size of it is 32 bits. This is one of the truncated version.

4. SHA512 : This hash function belong to hash class SHA-2, the internal block size of it is 64 bits.

5. SHA1 : The 160 bit hash function that resembles MD5 hash in working and was discontinued to be used seeing its security vulnerabilities.

In [28]:
import hashlib


### SHA256

In [32]:
str1="Sumar Fawaz"
### Process of Hashing
### First step is to encode
### Then Apply hashing Algorithms
hashedval = hashlib.sha256(str1.encode()) 
print(hashedval)

<sha256 _hashlib.HASH object @ 0x00000203DC49F350>


In [34]:
###convert this value to hexadecimal
hashedval.hexdigest()

'f7cbe5e99b52ed6e10841f049d5701a4ba2e7599677ec8645d9be35ef314ebb0'

### SHA384

In [41]:
hashedval=hashlib.sha384(str1.encode()) 
hashedval.hexdigest()

'7468cb7c496eff00167682113e1ed2ee9accb423780f31723199d924c07771b3e0712bd32691795178a90c0b337245bf'

## SHA224


In [44]:
hashedval=hashlib.sha224(str1.encode()) 
hashedval.hexdigest()

'6bca5b435c24bdac3793df4e45331e29cbd96a9b42c87b2209d5a953'

## SHA512

In [47]:
hashedval=hashlib.sha512(str1.encode()) 
hashedval.hexdigest()

'0a187b278b505db626f99e78adfe53f458ae67a0e3b6a8fb06520ed76b5e09dbce5700ad62e5659b51dfa86a37e977b509f88ebe5ee8d3f67f74891c3445c0b7'

In [49]:
hashedval=hashlib.sha1(str1.encode()) 
hashedval.hexdigest()

'f9a99189861570cb046a9b45c44a4ee7544474dc'

# Public, Protected, and Private access modifiers

In [52]:
### All the class variables are public
class Car():
    def __init__(self,windows,doors,enginetype):
        self.windows=windows
        self.doors=doors
        self.enginetype=enginetype

In [54]:
audi=Car(4,5,"Diesel")

In [56]:
audi.windows=5

In [66]:
### All the class variables are protected. can be only accessed using Subclasses
### They are declared using only one underscore
class Car():
    def __init__(self,windows,doors,enginetype):
        self._windows=windows
        self._doors=doors
        self._enginetype=enginetype

In [68]:
class Truck(Car):
    def __init__(self,windows,doors,enginetype,horsepower):
        super().__init__(windows,doors,enginetype)
        self.horsepowwer=horsepower

In [64]:
truck=Truck(4,4,"Diesel",4000)
dir(truck)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_doors',
 '_enginetype',
 '_windows',
 'horsepowwer']

In [70]:
truck._doors=5


In [72]:
truck._doors


5

In [76]:
### private variables can only be changed from that particular class but python just gives it 
### as an indication to the developer that it is private and can only be changed within that class
### it is declared using 2 underscores
class Car():
    def __init__(self,windows,doors,enginetype):
        self.__windows=windows
        self.__doors=doors
        self.__enginetype=enginetype

In [81]:
audi._Car__doors=5

In [85]:
audi._Car__doors
#still it can be changed because python doesn't impose strong restrictions

5

In [87]:
dir(audi)


['_Car__doors',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'doors',
 'enginetype',
 'windows']